Add Gateway API v1.1 Blog Post (#45971)
kubernetes/website: content/en/blog/_posts/2024-05-13-gateway-api-v1.1/index.md

---
layout: blog
title: "Gateway API v1.1: Service mesh, GRPCRoute, and a whole lot more"
date: 2024-05-13T19:00:00-08:00
slug: gateway-api-v1-1
author: >

>   Richard Belleville (Google),
>   Frank Budinsky (IBM),
>   Arko Dasgupta (Tetrate),
>   Flynn (Buoyant),
>   Candace Holman (Red Hat),
>   John Howard (Solo.io),
>   Christine Kim (Isovalent),
>   Mattia Lavacca (Kong),
>   Keith Mattix (Microsoft),
>   Mike Morris (Microsoft),
>   Rob Scott (Google),
>   Grant Spence (Red Hat),
>   Gina Yeh (Google),
>   and other review and release note contributors.

Reviewer: How about:
```suggestion
  Gateway API 1.1 release team
```
or the same, but with a link to the team. See https://kubernetes.io/blog/2024/04/17/kubernetes-v1-30-release/ for how we do that.

---

> ![Gateway API Logo](gateway-api-logo.svg)

Author: ok

and GRPCRoute for route-level configuration. The BackendLBPolicy and route-level
APIs provide the same session persistence configuration, including session
timeouts, session name, session type, and cookie lifetime type.

Below is an example configuration of `BackendLBPolicy` that enables cookie-based
session persistence for the `foo` service. It sets the session name to
`foo-session`, defines absolute and idle timeouts, and configures the cookie to
be a session cookie:

```
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendLBPolicy
metadata:
  name: lb-policy
  namespace: foo-ns
spec:
  targetRefs:
  - group: core
    kind: service
    name: foo
  sessionPersistence:
    sessionName: foo-session
    absoluteTimeout: 1h
    idleTimeout: 30m
    type: Cookie
    cookieConfig:
      lifetimeType: Session
```

### Everything else

#### [TLS Terminology Clarifications](https://gateway-api.sigs.k8s.io/geps/gep-2907/)

As part of a broader goal of making our TLS terminology more consistent
throughout the API, we've introduced some breaking changes to BackendTLSPolicy.
This has resulted in a new API version (v1alpha3) and will require any existing

> users of this policy to uninstall the v1alpha2 version before installing this

Reviewer: Potentially:
```suggestion
users of this policy to **uninstall** the v1alpha2 version, before you install this
```

I'm inferring people also need to delete their BackendTLSPolicies before they delete the CRD; have I got that right? If so, let's be more explicit about that.

Reviewer 2: > I'm inferring people also need to delete their BackendTLSPolicies before they delete the CRD; have I got that right? 

No need, deleting the CRD will also delete any CRs based on it.

Reviewer: What I'm indirectly wondering: could there be a finalizer (baked in, not something customizer) that prevents or delays removal of a particular BackendTLSPolicy? If so, I would call that out here.

Reviewer 3: So here's the (more) complete story. (Note that in the below, I'm using "CRD" to mean literally the CustomResourceDefinition resource that tells the API server how to handle custom resources, and "CR" to refer to instances of the custom resources thus defined. We often use those interchangeably, but it matters here.)

- If you just straight up delete the BackendTLSPolicy CRD, the API server will instantly and silently remove any BackendTLSPolicy CRs from the cluster, with no confirmation needed and no quarter given.
- If, on the other hand, your cluster has a `v1alpha2` BackendTLSPolicy CRD in place, and you attempt to apply a BackendTLSPolicy that no longer serves `v1alpha2` but instead serves `v1alpha3`, there is an interlock in the API server that will not accept the new CRD until you have explicitly told the API server that no `v1alpha2` CRs are in use. (You can lie to it, of course, and tell it that you're OK deleting `v1alpha2` when it's really not, but in that case on your own head be it.)

We haven't included the above in the docs because:

- the upshot is "installing `v1alpha3` requires you to take explicit action to tell the system that it's OK", which we _do_ include in the docs;
- _which_ explicit action you take depends on how you want to approach things; and
- we had previously thought that this was pretty well covered by the existing CRD documentation.

So, question back to you: do you think we're wrong there and we need to go into it more? or possibly write something else that goes into more detail, to which we then could link?

Reviewer: I don't think the proposed text is wrong, but I did want to check that you think this particular detail looks right.

Some controllers set finalizers that make the story more complicated (imagine if a BackendTLSPolicy could have a finalizer, and this might indirectly block removal of the CRD). It doesn't sound like that's the case here.

Reviewer 3: 🤦‍♂️ Ugh, sorry for misunderstanding and explaining what you already knew, then. Correct – finalizers shouldn't be in play for BackendTLSPolicy.

Reviewer: (no worries!)

Author: At the moment there are only one or two implementations of the experimental BackendTLSPolicy and so the meaning of "users" in this context arguably refers to implementations.  I will change this to:

> This has resulted in a new API version (v1alpha3) and will require any existing
> implementations of this policy to properly handle the version upgrade, e.g.
> by backing up data and uninstalling the v1alpha2 version before installing this
> newer version.

---
layout: blog
title: "Gateway API v1.1: Service mesh, GRPCRoute, and a whole lot more"
date: 2024-05-13T19:00:00-08:00
slug: gateway-api-v1-1
author: >
  [Richard Belleville](https://github.com/gnossen) (Google),
  [Frank Budinsky](https://github.com/frankbu) (IBM),
  [Arko Dasgupta](https://github.com/arkodg) (Tetrate),
  [Flynn](https://github.com/kflynn) (Buoyant),
  [Candace Holman](https://github.com/candita) (Red Hat),
  [John Howard](https://github.com/howardjohn) (Solo.io),
  [Christine Kim](https://github.com/xtineskim) (Isovalent),
  [Mattia Lavacca](https://github.com/mlavacca) (Kong),
  [Keith Mattix](https://github.com/keithmattix) (Microsoft),
  [Mike Morris](https://github.com/mikemorris) (Microsoft),
  [Rob Scott](https://github.com/robscott) (Google),
  [Grant Spence](https://github.com/gcs278) (Red Hat),
  [Shane Utt](https://github.com/shaneutt) (Kong),
  [Gina Yeh](https://github.com/ginayeh) (Google),

>   and other review and release note contributors.

Reviewer: ```suggestion
  and other review and release note contributors
```

If we can make this smaller, it'll help I think. Some people will stop reading before they reach the end of this list.

Author: I'd like to keep it as is for now.  Next time, we'll talk about this in our community gathering prior to creating the PR.

kind: HTTPRoute
metadata:
  name: color-canary
  namespace: faces
spec:
  parentRefs:
    - name: color
      kind: Service
      group: ""
      port: 80
  rules:
  - backendRefs:
    - name: color
      port: 80
      weight: 50
    - name: color2
      port: 80
      weight: 50
```
      
This would split traffic sent to the `color` Service in the `faces` namespace
50/50 between the original `color` Service and the `color2` Service, using a
portable configuration that's easy to move from one mesh to another.

#### [GRPCRoute](https://gateway-api.sigs.k8s.io/guides/grpc-routing/)
If you are already using the experimental version of GRPCRoute, we recommend holding
off on upgrading to the standard channel version of GRPCRoute until the
controllers you're using have been updated to support GRPCRoute v1. Until then,
it is safe to upgrade to the experimental channel version of GRPCRoute in v1.1
that includes both v1alpha2 and v1 API versions.

#### [ParentReference Port](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io%2fv1.ParentReference)
The `port` field was added to ParentReference, allowing you to attach resources

> to Gateway Listeners, Services (experimental), or other parent resources

Reviewer 4: Is binding to Service still experimental now that Service Mesh Support is promoted to standard?
```suggestion
to Gateway Listeners, Services, or other parent resources
```

---
layout: blog

> title: "Gateway API v1.1: Feature Graduation and Other Updates"

Reviewer 2: Completely skipped over this section in the initial review, but maybe we should include some of the headline features in the title. Also considered session persistence of client cert verification but either of those lead to a pretty long title.

```suggestion
title: "Gateway API v1.1: Service Mesh, GRPCRoute, and a whole lot more"
```

date: 2024-05-13T19:00:00-08:00
slug: gateway-api-v1-1
author: >
  Richard Belleville (Google),
  Frank Budinsky (IBM),
  Arko Dasgupta (Tetrate),
  Flynn (Buoyant),
  Candace Holman (Red Hat),
  John Howard (Solo.io),
  Mattia Lavacca (Kong),

>   Christine Kim (Isovalent),

Reviewer 2: Nit: If you're trying to order these names alphabetically be last name,I think these need to be swapped.

  Grant Spence (Red Hat),
  Gina Yeh (Google),
  and other reviewer and release note contributors.
---

![Gateway API Logo](gateway-api-logo.svg)

> In the wake of the GA release of Gateway API last October, the Kubernetes
> SIG Network buoyantly announces the v1.1 release of

Reviewer 2: Nit: Suggest removing "the" and finding another word in place of "buoyantly"
```suggestion
In the wake of the GA release of Gateway API last October, Kubernetes
SIG Network happily announces the v1.1 release of
```

> [Gateway API](https://gateway-api.sigs.k8s.io/). Gateway API features for
> **Gateway API for Service Mesh (GAMMA)**, **GRPCRoute**, **ParentReference Port**,
> and **Conformance Profiles and Reports** will graduate to Standard.
> In addition, **Gateway Client Certificate Verification** and **Session
> Persistence** are adopted as new, Experimental features. Clarifications to TLS
> terminology and gateway merging are added to strengthen the release.

Reviewer 2: Maybe we should just highlight the biggest new features at this point and include more details below?
```suggestion
[Gateway API](https://gateway-api.sigs.k8s.io/). In this release, several features
are graduating to Standard Channel (GA), notably including support for Service Mesh,
and GRPCRoute. We're also introducing some new experimental features, including
Session Persistence and Client Certificate Verification. Read on for the full list of 
what's included in this release of Gateway API.
```

Reviewer 5: Nit: should "service mesh" be lowercase in body copy?

Author: All suggestions added.

## What's new

### Graduation to Standard

> This release includes the graduation to _Standard_ of four eagerly awaited features.

Reviewer 2: ```suggestion
This release includes the graduation to _Standard Channel_ of four eagerly awaited features.
```

This means they are no longer experimental concepts. Inclusion in the Standard

> release channel denotes a high
> level of confidence in the API surface and provides guarantees of backward
> compatibility.  Though these features are now considered stable, that does not

Reviewer 2: ```suggestion
release channel is equivalent to General Availability and comes with strong backwards
compatibility guarantees.  Though these features are now considered stable, that does not
```

> mean that they are complete. These and other features will continue to evolve via
> the Experimental channel. For more information on how all of this works, refer

Reviewer 2: Not quite sure how to say this, but I'm worried that if we mention experimental channel here we may give the notion that unstable changes may happen to these features or that new features will only ever be available in experimental channel. Trying to avoid either of those potential points of confusion, but open to alternatives here.
```suggestion
mean that they are complete. In the future, backwards compatible additions may be made
such as new features or supported values. For more information on how all of this works, refer
```

Reviewer 3: Maybe

> This release includes the graduation to Standard of four eagerly awaited features. This means they are no longer experimental concepts; inclusion in the Standard release channel denotes a high level of confidence in the API surface and provides guarantees of backward compatibility. Of course, as with any other Kubernetes API, Standard channel features can continue to evolve with backward-compatible additions over time, and we certainly expect further refinements and improvements to these new features in the future. For more information on how all of this works, refer to the [Gateway API Versioning Policy](https://gateway-api.sigs.k8s.io/concepts/versioning/).

?

Reviewer 2: That's great @kflynn, thanks! Let's go with that

to the [Gateway API Versioning Policy](https://gateway-api.sigs.k8s.io/concepts/versioning/).

> #### [GAMMA](https://gateway-api.sigs.k8s.io/mesh/)

Reviewer 2: ```suggestion
#### [Service Mesh Support](https://gateway-api.sigs.k8s.io/mesh/)
```

Reviewer 2: This may just be my own bias showing through, but I think we've generally been moving away from GAMMA terminology in favor of "Service Mesh support". Ultimately will defer to preference of @howardjohn, @keithmattix, @kflynn, or @mikemorris on this one though.

Reviewer 6: Strong +1

Reviewer 3: Yeah, I agree. We use GAMMA as synecdoche a lot, but "Service Mesh Support" is better for public consumption.

Reviewer 5: +1, it's been easier to understand for new audiences when I've been explaining this as "Gateway API for service mesh"

Reviewer 7: +1 on service mesh instead of GAMMA

Author: Done.

API to manage ingress traffic and mesh traffic, reusing the same policy and
routing interfaces. In Gateway API v1.1, routes (such as HTTPRoute) can now have
a Service as a `parentRef`, to control how traffic to specific services behave.
For more information, read the 
[Gateway API service mesh documentation](https://gateway-api.sigs.k8s.io/mesh/)
or see the
[list of Gateway API implementations
](https://gateway-api.sigs.k8s.io/implementations/#service-mesh-implementation-status).

As an example, one could do a canary deployment of a workload deep in an
application's call graph with an HTTPRoute as follows:

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: color-canary
  namespace: faces
spec:
  parentRefs:
    - name: color
      kind: Service
      group: core
      port: 80
  rules:
  - backendRefs:
    - name: color
      port: 80
      weight: 50
    - name: color2
      port: 80
      weight: 50
```
      
This would split traffic sent to the `color` Service in the `faces` namespace
50/50 between the original `color` Service and the `color2` Service, using a
portable configuration that's easy to move from one mesh to another.

> Leading Contributors: Flynn, John Howard, Keith Mattix, Mike Morris

Reviewer 2: Complete optional, but could be a nice touch to link to GitHub profiles

Reviewer 5: ```suggestion
Leading Contributors: [Flynn](https://github.com/kflynn), [John Howard](https://github.com/howardjohn/), [Keith Mattix](https://github.com/keithmattix/), [Mike Morris](https://github.com/mikemorris/)
```

Author: The GitHub profile are listed in the Release Notes already, so I'd iike to call this done with my last round of changes.

Reviewer 2: I agree that this should not block the PR, but if someone has the time to add suggestions for all names throughout the PR, I think that would be a helpful addition, I'm not sure how many people will bother to click through to the release notes after reading this.

Author: I've decided to remove the Leading Contributors lines.  There are no other Gateway API blog articles where we list contributors, and I would prefer to highlight the team as a whole.  No hard feelings felt by anyone, I hope.

Author: Since I have to make other changes, I will add the github profile links to the authors list.  This seems to be the way others are handling the attributions.

`foo-session`, defines absolute and idle timeouts, and configures the cookie to
be a session cookie:

```
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: BackendLBPolicy
metadata:
  name: lb-policy
  namespace: foo-ns
spec:
  targetRefs:
  - group: core
    kind: service
    name: foo
  sessionPersistence:
    sessionName: foo-session
    absoluteTimeout: 1h
    idleTimeout: 30m
    type: Cookie
    cookieConfig:
      lifetimeType: Session
```

Leading Contributors: Grant Spence, Gina Yeh

#### [TLS Terminology Clarifications](https://gateway-api.sigs.k8s.io/geps/gep-2907/)

As part of a broader goal of making our TLS terminology more consistent
throughout the API, we've introduced some breaking changes to BackendTLSPolicy.
This has resulted in a new API version (v1alpha3) and will require any existing
users of this policy to uninstall the v1alpha2 version before installing this
newer version.

Any references to v1alpha2 BackendTLSPolicy fields will need to be updated to
v1alpha3. Specific changes to fields include:
- `targetRef` becomes `targetRefs` to allow a BackendTLSPolicy to attach to
multiple targets
- `tls` becomes `validation`

> - The type name BackendTLSPolicyConfig becomes BackendTLSPolicyValidation

Reviewer 2: Recommend go type changes out (same comment below)

- `tls.caCertRefs` becomes `validation.caCertificateRefs`
- `tls.wellKnownCACerts` becomes `validation.wellKnownCACertificates`
- The type name WellKnownCACertType becomes WellKnownCACertificatesType
and

> - The constant WellKnownCACertSystem becomes WellKnownCACertificatesSystem

Reviewer 2: I'd recommend leaving out any reference to go types since our primary audience for this blog will be people using the APIs directly with something like YAML, not Go types.

Reviewer 3: I could totally get behind this, FTR. 👍

Author: Done.

slug: gateway-api-v1-1
author: >
  Richard Belleville (Google),
  Frank Budinsky (IBM),
  Arko Dasgupta (Tetrate),
  Flynn (Buoyant),
  Candace Holman (Red Hat),
  John Howard (Solo.io),
  Christine Kim (Isovalent),
  Mattia Lavacca (Kong),
  Grant Spence (Red Hat),
  Gina Yeh (Google),
  and other reviewer and release note contributors.
---

![Gateway API Logo](gateway-api-logo.svg)

Following the GA release of Gateway API last October, Kubernetes
SIG Network is pleased to announce the v1.1 release of
[Gateway API](https://gateway-api.sigs.k8s.io/). In this release, several features are graduating to
_Standard Channel_ (GA), notably including support for service mesh and
GRPCRoute. We're also introducing some new experimental features, including
session persistence and client certificate verification. Read on for the full
list of what's included in this release of Gateway API.

**Service Mesh Support**, **GRPCRoute**, **ParentReference Port**,
and **Conformance Profiles and Reports** will graduate to Standard.
In addition, **Gateway Client Certificate Verification** and **Session
Persistence and BackendLBPolicy** are adopted as new, Experimental features.

> Clarifications to TLS terminology is added to strengthen the release.

Reviewer 2: Recommend leaving this level of detail out of the intro in favor of the higher level intro paragraph above.

provides guarantees of backward compatibility. Of course, as with any other
Kubernetes API, Standard Channel features can continue to evolve with
backward-compatible additions over time, and we certainly expect further
refinements and improvements to these new features in the future.
For more information on how all of this works, refer to the
[Gateway API Versioning Policy](https://gateway-api.sigs.k8s.io/concepts/versioning/).

#### [Service Mesh Support](https://gateway-api.sigs.k8s.io/mesh/)
Service mesh support in Gateway API allows service mesh users to use the same
API to manage ingress traffic and mesh traffic, reusing the same policy and
routing interfaces. In Gateway API v1.1, routes (such as HTTPRoute) can now have
a Service as a `parentRef`, to control how traffic to specific services behave.
For more information, read the 
[Gateway API service mesh documentation](https://gateway-api.sigs.k8s.io/mesh/)
or see the
[list of Gateway API implementations
](https://gateway-api.sigs.k8s.io/implementations/#service-mesh-implementation-status).

As an example, one could do a canary deployment of a workload deep in an
application's call graph with an HTTPRoute as follows:

```

> apiVersion: gateway.networking.k8s.io/v1beta1

Reviewer 2: ```suggestion
apiVersion: gateway.networking.k8s.io/v1
```

kind: HTTPRoute
metadata:
  name: color-canary
  namespace: faces
spec:
  parentRefs:
    - name: color
      kind: Service

>       group: core

Reviewer 2: ```suggestion
      group: ""
```

    kind: service
    name: foo
  sessionPersistence:
    sessionName: foo-session
    absoluteTimeout: 1h
    idleTimeout: 30m
    type: Cookie
    cookieConfig:
      lifetimeType: Session
```

Leading Contributors: Grant Spence, Gina Yeh

### Everything else

#### [TLS Terminology Clarifications](https://gateway-api.sigs.k8s.io/geps/gep-2907/)

As part of a broader goal of making our TLS terminology more consistent
throughout the API, we've introduced some breaking changes to BackendTLSPolicy.
This has resulted in a new API version (v1alpha3) and will require any existing
users of this policy to uninstall the v1alpha2 version before installing this
newer version.

Any references to v1alpha2 BackendTLSPolicy fields will need to be updated to
v1alpha3. Specific changes to fields include:
- `targetRef` becomes `targetRefs` to allow a BackendTLSPolicy to attach to
multiple targets
- `tls` becomes `validation`
- `tls.caCertRefs` becomes `validation.caCertificateRefs`
- `tls.wellKnownCACerts` becomes `validation.wellKnownCACertificates`

Leading Contributors: Candace Holman, Rob Scott

For a full list of the changes included in this release, please refer to the
[v1.1.0 release notes.

> ](https://github.com/kubernetes-sigs/gateway-api/releases)

Reviewer 2: ```suggestion
](https://github.com/kubernetes-sigs/gateway-api/releases/tag/v1.1.0)
```
